### PR TITLE
i_900: add support for multiple groups per team

### PR DESCRIPTION
**Pull request by johnbrvc**
### Description of what the PR does
This PR provides built-in support for an account to belong to multiple groups, instead of just one.  The allows per-group display of standings, scoreboard generation and report generation.

The following changes were made to support multiple groups:

1. Add `HashSet<ElementId> groupIds` to `Account `object - this is a hash table containing all groups the account is a member of
2. Rename the old group id to `primaryGroupId` which is now the single **CMS** supplied group ID.  These is currently used by several unit tests, and the **DSA**/**NSA** for group ranking.  (Note: Group ranking has to be fixed so it ranks the team for each group the team is a member of.  This is a non-trivial change).
3. Add methods to `Account` to support adding a group to an account, fetching the groups for an account, clearing the groups.
4. Change PC2 API `ITeam`, `ITeamImplementation` to support multiple groups for the PC2 api (**WTI**) interface.
5. Change **NSA**  to add all groups to generated XML for standings
6. `PacketFormatter `shows all groups for an account.
7. In `ContestXML`, when generating contest xml, add all groups for each account
8. In `ExportAccounts`, include all groups for each account
9. Change `ICPCAccount` to support multiple groups, although currently, only one is ever present in an ICPC "_tab_" file.  Change made for consistency.
10. Change `ICPCImportData` to pass all groups when creating an `ICPCAccount`
11. Fix `Account `report generation to include all groups the team is a member of in the report
12. Add support in `ScoreboardUtilities` to support multiple groups for routines that previously only checked one group per account.
13. Add method `getGroupFilteredRuns()` to `ScoreboardUtilities` to return runs appropriate for selected groups only. (used by **DSA**/**NSA**)
14. Change `JSONTool` to generate multiple groups per account
15. In `PacketHandler`, change the routine that calculates which problems a team can see (it's now based on multiple groups, not just one)
16. Change `LoadAccounts` to accept a CSV list of groups on importing accounts instead of just a single group.
17. In `CLICSAwardUtilities`, change getGroupForTeam to getGroupsForTeam
18. In `InternalContest`, change `addProblem` to account for all groups the team belongs to
19. `ProblemGroupAssignmentReport` was changed to check all groups for a team.
20. In `EventFeedXML`, `EventFeedXML2013` and `ResolverEventFeedXML` add all groups a team is a member of to XML feed as separate group tags
21. In `TeamData`, use renamed `primaryGroupId` - it's ok to only support one group here since this is a legacy feed object
22. In `ICPCTSVLoader`, when processing `teams.tsv`, `teams2.tsv`, accept CSV list of groups in the group field.
23. Many copyright date changes to 2024
24. Modified `AccountsPane` and `AccountsTablePane` to show all groups a team is a member of in the **Groups** column.
25. Add support to `EditAccountPane` to allow the selection of multiple groups for an account using checkboxes (still dome code that is commented out that needs to be deleted before approval)
26. Change `ReviewAccountLoad` to show multiple groups correctly, and to detect changes properly in groups.
27. Many _JUnit_ tests changed to support and test multiple group support
28. In **DSA** and **NSA**, allow standings to be calculated for a given set of groups.  For now, maintain `divisionNumber `as well for WTI.
29. Update `StandingsTablePane` and `StandingsHTMLPane` to allow selection of which groups to show the standings for on thick client standings panes.
30. Update `ScoreboardView`, `ScoreboardModule` and related classes to deal with multiple groups.  Scoreboards are genereated for every group marked as "show on scoreboard".  Files are created with the group name embedded in the file name.
31. Added report filtering by group (useful for generating event feeds for specific groups)
32. Added event feed filtering by group as well (as a result of 31.)
33. Added streaming event feed filtering by queryparam: groupids=cmsid1,cmsid2,...
34. CI: added `cacerts.pc2` to `.gitignore`
35. CI: Various modules had extra spaces removed at ends of lines
36. CI: Various modules had `@overrides` added on inherited methods
37. CI: Eclipse removed dunsel typecasts in **NSA** module, `EventFeedXML`
38. CI: optimizations made - hash tables instead of linear searches
39. CI: Spelling errors in member variable names and method names fixed.

TODO:

1. Testing
2. Testing
3. Testing
4. Testing
5. Testing

### Issue which the PR addresses
Partial implementation #900 

### Environment in which the PR was developed (OS,IDE, Java version, etc.)
Windows 11
Java 1.8.0_321

### Precise steps for _testing_ the PR (i.e., how to demonstrate that it works correctly)
Where to start?
Everything that deals with accounts has to be tested:  `AccountTablePane`, Standings, HTML Standings, Editing accounts, Editing groups, scoreboards, reports, problems, ...
More to come on this once complete.

### CODE REVIEW

File: test/edu/csus/ecs/pc2/core/StringUtilitiesTest.java
Comment (troy2914) [inline test/edu/csus/ecs/pc2/core/StringUtilitiesTest.java:1]: update copyright year

File: test/edu/csus/ecs/pc2/core/imports/LoadAccountsTest.java
Comment (troy2914) [inline test/edu/csus/ecs/pc2/core/imports/LoadAccountsTest.java:1]: update copyright year here

File: test/edu/csus/ecs/pc2/core/list/AccountListTest.java
Comment (troy2914) [inline test/edu/csus/ecs/pc2/core/list/AccountListTest.java:1]: update copyright year

File: test/edu/csus/ecs/pc2/exports/ccs/ResolverEventFeedXMLTest.java
Comment (troy2914) [inline test/edu/csus/ecs/pc2/exports/ccs/ResolverEventFeedXMLTest.java:1]: update copyright year

File: test/edu/csus/ecs/pc2/util/ScoreboardVariableReplacerTest.java
Comment (troy2914) [inline test/edu/csus/ecs/pc2/util/ScoreboardVariableReplacerTest.java:1]: update copyright year.

File: test/edu/csus/ecs/pc2/exports/ccs/EventFeedXML2013Test.java
Comment (troy2914) [inline test/edu/csus/ecs/pc2/exports/ccs/EventFeedXML2013Test.java:1]: update copyright year

File: test/edu/csus/ecs/pc2/core/scoring/NewScoringAlgorithmTestLong.java
Comment (troy2914) [inline test/edu/csus/ecs/pc2/core/scoring/NewScoringAlgorithmTestLong.java:1]: update copyright year

File: test/edu/csus/ecs/pc2/core/scoring/DefaultScoringAlgorithmTest.java
Comment (troy2914) [inline test/edu/csus/ecs/pc2/core/scoring/DefaultScoringAlgorithmTest.java:1]: update copyright year

File: test/edu/csus/ecs/pc2/core/model/SampleContest.java
Comment (troy2914) [inline test/edu/csus/ecs/pc2/core/model/SampleContest.java:1]: update copyright year

File: test/edu/csus/ecs/pc2/core/model/InternalContestTest.java
Comment (troy2914) [inline test/edu/csus/ecs/pc2/core/model/InternalContestTest.java:1]: update copyright year

File: test/edu/csus/ecs/pc2/core/imports/clics/CLICSAwardUtilitiesTest.java
Comment (troy2914) [inline test/edu/csus/ecs/pc2/core/imports/clics/CLICSAwardUtilitiesTest.java:1]: update copyright year

File: src/edu/csus/ecs/pc2/ui/StandingsTablePane.java
Comment (troy2914) [inline src/edu/csus/ecs/pc2/ui/StandingsTablePane.java:1]: update copyright year

File: src/edu/csus/ecs/pc2/ui/EditAccountPane.java
Comment (troy2914) [inline src/edu/csus/ecs/pc2/ui/EditAccountPane.java:1]: imo delete the to be deleted old stuff.  vs just commenting it out.

**Review by troy2914 — CHANGES_REQUESTED** [review]
Generally code looks good,  the whole file source formatting is a pain to review.

**Comment by troy2914** [comment]
You also saw this build was failing?

File: src/edu/csus/ecs/pc2/core/model/InternalContest.java
[the remark this answers is not in the document]
Comment (johnbrvc) [reply src/edu/csus/ecs/pc2/core/model/InternalContest.java:1]: Fixed.

File: src/edu/csus/ecs/pc2/core/report/AccountsReport.java [hunk 1]
```diff
@@ -151,16 +152,27 @@ private void printAccount(PrintWriter printWriter, Account account) {
...
-            if (group != null) {
-                printWriter.print("group '"+group+"' ("+groupId+")");
-            } else {
-                printWriter.print("group invalid ("+groupId+")");
+        if(account.getGroupIds() != null) {
+            printWriter.print("groups '");
+            boolean first = true;
+            for(ElementId elementId: account.getGroupIds()) {
+                if(first == false) {
+                    printWriter.print(",");;
```
[the remark this answers is not in the document]
Comment (johnbrvc) [reply src/edu/csus/ecs/pc2/core/report/AccountsReport.java]: Fixed.

File: src/edu/csus/ecs/pc2/core/scoring/IScoringAlgorithm.java [hunk 2]
```diff
@@ -1,23 +1,25 @@
 // Copyright (C) 1989-2023 PC2 Development Team: John Clevenger, Douglas Lane, Samir Ashoo, and Troy Boudreau.
```
[the remark this answers is not in the document]
Comment (johnbrvc) [reply src/edu/csus/ecs/pc2/core/scoring/IScoringAlgorithm.java]: Fixed.

File: src/edu/csus/ecs/pc2/exports/ccs/StandingsJSON.java
[the remark this answers is not in the document]
Comment (johnbrvc) [reply src/edu/csus/ecs/pc2/exports/ccs/StandingsJSON.java:1]: Fixed.

File: src/edu/csus/ecs/pc2/exports/ccs/TeamsJSON.java [hunk 3]
```diff
@@ -8,22 +8,23 @@
...
 import edu.csus.ecs.pc2.core.model.ElementId;
 import edu.csus.ecs.pc2.core.model.Group;
 import edu.csus.ecs.pc2.core.model.IInternalContest;
-import edu.csus.ecs.pc2.core.model.ClientType.Type;
 
 /**
  * Team information in CLI 2016 JSON format.
- * 
+ *
+ * TODO: Remove this class - it is not refereneced by any module. -- JB
```
[the remark this answers is not in the document]
Comment (johnbrvc) [reply src/edu/csus/ecs/pc2/exports/ccs/TeamsJSON.java]: Deleted.

File: test/edu/csus/ecs/pc2/core/model/SampleContest.java
Comment (johnbrvc) [reply test/edu/csus/ecs/pc2/core/model/SampleContest.java:1]: Fixed.

File: test/edu/csus/ecs/pc2/core/model/InternalContestTest.java
Comment (johnbrvc) [reply test/edu/csus/ecs/pc2/core/model/InternalContestTest.java:1]: Fixed.

File: test/edu/csus/ecs/pc2/core/imports/clics/CLICSAwardUtilitiesTest.java
Comment (johnbrvc) [reply test/edu/csus/ecs/pc2/core/imports/clics/CLICSAwardUtilitiesTest.java:1]: Fixed.

File: src/edu/csus/ecs/pc2/ui/StandingsTablePane.java
Comment (johnbrvc) [reply src/edu/csus/ecs/pc2/ui/StandingsTablePane.java:1]: Fixed.

File: src/edu/csus/ecs/pc2/ui/EditAccountPane.java
Comment (johnbrvc) [reply src/edu/csus/ecs/pc2/ui/EditAccountPane.java:1]: Deleted.

File: src/edu/csus/ecs/pc2/ui/ICPCAccountPane.java [hunk 4]
```diff
@@ -1,15 +1,27 @@
 // Copyright (C) 1989-2019 PC2 Development Team: John Clevenger, Douglas Lane, Samir Ashoo, and Troy Boudreau.
```
[the remark this answers is not in the document]
Comment (johnbrvc) [reply src/edu/csus/ecs/pc2/ui/ICPCAccountPane.java]: Fixed.

File: src/edu/csus/ecs/pc2/ui/ICPCLoadPane.java [hunk 5]
```diff
@@ -457,16 +463,18 @@ private void info(String message) {
 
     /**
      * Account Listener Implementation.
-     * 
+     *
      * @author pc2@ecs.csus.edu
-     * @version $Id: ICPCPane.java 2275 2010-11-30 03:39:24Z laned $
+     * @version $Id: ICPCLoadPane.java 2275 2010-11-30 03:39:24Z laned $
```
[the remark this answers is not in the document]
Comment (johnbrvc) [reply src/edu/csus/ecs/pc2/ui/ICPCLoadPane.java]: Removed.

File: src/edu/csus/ecs/pc2/ui/ICPCLoadPane.java [hunk 6]
```diff
@@ -522,26 +536,30 @@ public void run() {
 
     /**
      * Group Listener for ICPC Pane.
-     * 
+     *
      * @author pc2@ecs.csus.edu
-     * @version $Id: ICPCPane.java 2275 2010-11-30 03:39:24Z laned $
+     * @version $Id: ICPCLoadPane.java 2275 2010-11-30 03:39:24Z laned $
```
[the remark this answers is not in the document]
Comment (johnbrvc) [reply src/edu/csus/ecs/pc2/ui/ICPCLoadPane.java]: Removed.

File: src/edu/csus/ecs/pc2/util/ScoreboardVariableReplacer.java
[the remark this answers is not in the document]
Comment (johnbrvc) [reply src/edu/csus/ecs/pc2/util/ScoreboardVariableReplacer.java:1]: Fixed.

File: test/edu/csus/ecs/pc2/api/implementation/TeamImplementationTest.java
[the remark this answers is not in the document]
Comment (johnbrvc) [reply test/edu/csus/ecs/pc2/api/implementation/TeamImplementationTest.java:1]: Fixed.

File: test/edu/csus/ecs/pc2/core/PacketHandlerTest.java
[the remark this answers is not in the document]
Comment (johnbrvc) [reply test/edu/csus/ecs/pc2/core/PacketHandlerTest.java:1]: Fixed.

File: test/edu/csus/ecs/pc2/core/StringUtilitiesTest.java
Comment (johnbrvc) [reply test/edu/csus/ecs/pc2/core/StringUtilitiesTest.java:1]: Fixed.

File: test/edu/csus/ecs/pc2/core/imports/LoadAccountsTest.java
Comment (johnbrvc) [reply test/edu/csus/ecs/pc2/core/imports/LoadAccountsTest.java:1]: Fixed.

File: test/edu/csus/ecs/pc2/core/list/AccountListTest.java
Comment (johnbrvc) [reply test/edu/csus/ecs/pc2/core/list/AccountListTest.java:1]: Fixed.

File: test/edu/csus/ecs/pc2/exports/ccs/ResolverEventFeedXMLTest.java
Comment (johnbrvc) [reply test/edu/csus/ecs/pc2/exports/ccs/ResolverEventFeedXMLTest.java:1]: Fixed.

File: test/edu/csus/ecs/pc2/util/ScoreboardVariableReplacerTest.java
Comment (johnbrvc) [reply test/edu/csus/ecs/pc2/util/ScoreboardVariableReplacerTest.java:1]: Fixed.

File: test/edu/csus/ecs/pc2/exports/ccs/EventFeedXML2013Test.java
Comment (johnbrvc) [reply test/edu/csus/ecs/pc2/exports/ccs/EventFeedXML2013Test.java:1]: Fixed.

File: test/edu/csus/ecs/pc2/core/scoring/NewScoringAlgorithmTestLong.java
Comment (johnbrvc) [reply test/edu/csus/ecs/pc2/core/scoring/NewScoringAlgorithmTestLong.java:1]: Fixed.

File: test/edu/csus/ecs/pc2/core/scoring/DefaultScoringAlgorithmTest.java
Comment (johnbrvc) [reply test/edu/csus/ecs/pc2/core/scoring/DefaultScoringAlgorithmTest.java:1]: Fixed.

**Review by troy2914 — APPROVED** [review]
What's here so far seems okay.  And build is succeeding now.

**Review by troy2914 — APPROVED** [review]
These changes look good to me.

File: src/edu/csus/ecs/pc2/core/model/Filter.java [hunk 7]
```diff
@@ -1,54 +1,59 @@
-// Copyright (C) 1989-2019 PC2 Development Team: John Clevenger, Douglas Lane, Samir Ashoo, and Troy Boudreau.
+// Copyright (C) 1989-2024 PC2 Development Team: John Clevenger, Douglas Lane, Samir Ashoo, and Troy Boudreau.
 package edu.csus.ecs.pc2.core.model;
 
 import java.io.Serializable;
+import java.util.Collection;
+import java.util.Collections;
```
Comment (troy2914) [inline src/edu/csus/ecs/pc2/core/model/Filter.java:6]: I'm confused by these?  I don't see them used?

File: src/edu/csus/ecs/pc2/core/report/EventFeedJSONReport.java
Comment (troy2914) [inline src/edu/csus/ecs/pc2/core/report/EventFeedJSONReport.java:1]: update year in copyright

File: src/edu/csus/ecs/pc2/ui/EditFilterPane.java
Comment (troy2914) [inline src/edu/csus/ecs/pc2/ui/EditFilterPane.java:1]: update copyright year please

File: src/edu/csus/ecs/pc2/ui/EditFilterPane.java [hunk 8]
```diff
@@ -319,10 +336,27 @@ private JPanel getTeamsPane() {
         }
         return teamsPane;
     }
-    
+
+    /**
+     * This method initializes teamFrame
```
Comment (troy2914) [inline src/edu/csus/ecs/pc2/ui/EditFilterPane.java:341]: teamFrame or a groupPane?

File: src/edu/csus/ecs/pc2/ui/ReportPane.java [hunk 9]
```diff
@@ -1,4 +1,3 @@
-// Copyright (C) 1989-2023 PC2 Development Team: John Clevenger, Douglas Lane, Samir Ashoo, and Troy Boudreau.
```
Comment (troy2914) [inline src/edu/csus/ecs/pc2/ui/ReportPane.java:1]: why remove, verses updating the year?

**Review by troy2914 — CHANGES_REQUESTED** [review]
getting closer.

**Review by troy2914 — COMMENTED** [review]
These changes seem reasonable,  but did not address any of my complaints about the previous commits.

File: src/edu/csus/ecs/pc2/core/model/Filter.java (hunk 7 above)
Comment (johnbrvc) [reply src/edu/csus/ecs/pc2/core/model/Filter.java:6]: HashSet requires them.  Automatically added by Eclipse and can't be removed.

File: src/edu/csus/ecs/pc2/core/report/EventFeedJSONReport.java
Comment (johnbrvc) [reply src/edu/csus/ecs/pc2/core/report/EventFeedJSONReport.java:1]: Fixed.
 [Commit](https://github.com/pc2ccs/pc2v9/pull/909/commits/ac4635f11971b2ebf29eb6f764a50d615603b0f4)

File: src/edu/csus/ecs/pc2/ui/EditFilterPane.java
Comment (johnbrvc) [reply src/edu/csus/ecs/pc2/ui/EditFilterPane.java:1]: Fixed. 
 [Commit](https://github.com/pc2ccs/pc2v9/pull/909/commits/ac4635f11971b2ebf29eb6f764a50d615603b0f4)

File: src/edu/csus/ecs/pc2/ui/EditFilterPane.java (hunk 8 above)
Comment (johnbrvc) [reply src/edu/csus/ecs/pc2/ui/EditFilterPane.java]: Yeah.  Fixed that one and a couple of other "legacy" cut&paste ones as well.
 [Commit](https://github.com/pc2ccs/pc2v9/pull/909/commits/ac4635f11971b2ebf29eb6f764a50d615603b0f4)

File: src/edu/csus/ecs/pc2/ui/ReportPane.java (hunk 9 above)
Comment (johnbrvc) [reply src/edu/csus/ecs/pc2/ui/ReportPane.java:1]: Why indeed?  Mistake.  Fixed. [Commit](https://github.com/pc2ccs/pc2v9/pull/909/commits/ac4635f11971b2ebf29eb6f764a50d615603b0f4)

**Review by troy2914 — APPROVED** [review]
looks good to me.

**Comment by troy2914** [comment]
just lookin at the results_D1.xml,  I think we need to add a attribute to the standingsHeader to show what the xml is showing..   like group_D1 or something.

**Comment by troy2914** [comment]
> just lookin at the results_D1.xml, I think we need to add a attribute to the standingsHeader to show what the xml is showing.. like group_D1 or something.

we discussed this and opted the included=1 on the grouplist/group

**Review by troy2914 — APPROVED** [review]
new changes seem reasonable to me.

**Comment by johnbrvc** [comment]
From @SamanwaySadhu on slack:

  [9:50 PM](https://icpcgreaterne-fak9235.slack.com/archives/D05JZG38PK8/p1707447035438329)
I've already gone through the code changes for the i900 branch and it looks good to me

**Pull request merged**